text/010-wasm-pack-watch-and-serve.md

```diff
@@ -0,0 +1,230 @@
+changes. Initially, we will just watch the crate directory for
+changes. Eventually, we can use [`cargo build --build-plan`][build-plan] to get
+a list of files that we should be watching.
+
+[notify]: https://crates.io/crates/notify
+[build-plan]: https://github.com/rust-lang/cargo/issues/5579
+
+## `wasm-pack serve`
+
+The `serve` subcommand starts a local Web server in the crate directory, watches
+for file changes, and re-builds the crate on changes.
+
+### Command Line Interface
+
+The `serve` subcommand takes a superset of the arguments that `wasm-pack build`
+takes. Like other `wasm-pack` subcommands, it takes an optional path to the
+crate directory as a positional argument, but if that is missing, then it
+defaults to the current directory. It has the usual flags for controlling the
+build profile, and the target environment. It takes extra options after `--`
+that get passed through straight to `cargo build`.
+
+Additionally, it has a `--no-watch` flag to disable watching the crate for
+changes to kick off automatic re-builds, and a `--port` option to specifiy the
+port the local server should bind to.
+
+```
+USAGE:
+    wasm-pack serve [FLAGS] [OPTIONS] [path] [-- <extra_options>...]
+
+FLAGS:
+        --dev              Create a development build. Enable debug info, and disable optimizations.
+        --no-typescript    By default a *.d.ts file is generated for the generated JS file, but this flag will disable
+                           generating this TypeScript file.
+    -h, --help             Prints help information
+        --profiling        Create a profiling build. Enable optimizations and debug info.
+        --release          Create a release build. Enable optimizations and disable debug info.
+    -V, --version          Prints version information
+        --no-watch         Do not watch the crate for changes to automaticaly rebuild.
+
+OPTIONS:
+    -m, --mode <mode>          Sets steps to be run. [possible values: no-install, normal, force] [default: normal]
+    -d, --out-dir <out_dir>    Sets the output directory with a relative path. [default: pkg]
+    -s, --scope <scope>        The npm scope to use in package.json, if any.
+    -t, --target <target>      Sets the target environment. [possible values: browser, nodejs, no-modules] [default:
+                               browser]
+    -p, --port <port>          Bind to this port number with the local development HTTP server. [default: 8000]
+
+ARGS:
+    <path>                The path to the Rust crate.
+    <extra_options>...    List of extra options to pass to `cargo build`
+```
+
+### Implementation
+
+Rather than integrating an HTTP server into the `wasm-pack` binary itself, we
+should leverage its ability to download and run other binaries. Exactly *which*
+local HTTP server binary is left as an unresolved question.
+
+The subcommand will start the local HTTP server and then execute the `watch`
+subcommand (unless `--no-watch` is supplied).
+
```

Comment:
What do you think about adding a `--root` option to the `wasm-pack serve` subcommand? The HTTP server will presumably need to figure out a filesystem root path to start hosting static files from, and while not specified here the cwd seems like a reasonable default, but having a `--root` option as well to configure it may be reasonable. For example if you customize `--out-dir` you may want to also customize `--root`, like:

```
wasm-pack serve --out-dir html/dist --root html
```

(or something like that)